frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1220 @@
+				.field("size", &self.size)
+				.field("top_items", &self.top_items)
+				.field("child_items", &self.child_items)
+				.finish()
+		}
+	}
+
+	impl<T: Config> Default for MigrationTask<T> {
+		fn default() -> Self {
+			Self {
+				current_top: Some(Default::default()),
+				current_child: Default::default(),
+				dyn_child_items: Default::default(),
+				dyn_top_items: Default::default(),
+				dyn_size: Default::default(),
+				prev_tick_child: Default::default(),
+				_ph: Default::default(),
+				size: Default::default(),
+				top_items: Default::default(),
+				child_items: Default::default(),
+			}
+		}
+	}
+
+	impl<T: Config> MigrationTask<T> {
+		/// Return true if the task is finished.
+		#[cfg(test)]
+		pub(crate) fn finished(&self) -> bool {
+			self.current_top.is_none() && self.current_child.is_none()
+		}
+
+		/// get the total number of keys affected by the current task.
+		pub(crate) fn dyn_total_items(&self) -> u32 {
+			self.dyn_child_items.saturating_add(self.dyn_top_items)
+		}
+
+		/// Migrate keys until either of the given limits are exhausted, or if no more top keys
+		/// exist.
+		///
+		/// Note that this returns after the **first** migration tick that causes exhaustion. In
+		/// other words, this should not be used in any environment where resources are strictly
+		/// bounded (e.g. a parachain), but it is acceptable otherwise (relay chain, offchain
+		/// workers).
+		pub(crate) fn migrate_until_exhaustion(&mut self, limits: MigrationLimits) {
+			log!(debug, "running migrations until {:?}", limits);
+			loop {
+				self.migrate_tick();
+				if self.exhausted(limits) {
+					break
+				}
+			}
+
+			// accumulate dynamic data into the storage items. 
+			self.size = self.size.saturating_add(self.dyn_size);
+			self.child_items = self.child_items.saturating_add(self.dyn_child_items);
+			self.top_items = self.top_items.saturating_add(self.dyn_top_items);
+			log!(debug, "finished with {:?}", self);
+		}
+
+		/// Check if there's any work left, or if we have exhausted the limits already.
+		fn exhausted(&self, limits: MigrationLimits) -> bool {
+			self.current_top.is_none() ||
+				self.dyn_total_items() >= limits.item ||
+				self.dyn_size >= limits.size
+		}
+
+		/// Migrate AT MOST ONE KEY. This can be either a top or a child key.
+		///
+		/// This function is the core of this entire pallet.
+		fn migrate_tick(&mut self) {
+			match (self.current_top.as_ref(), self.current_child.as_ref()) {
+				(Some(_), Some(_)) => {
+					// we're in the middle of doing work on a child tree.
+					self.migrate_child();
+				},
+				(Some(ref top_key), None) => {
+					if top_key.starts_with(DEFAULT_CHILD_STORAGE_KEY_PREFIX) &&
+						!self.prev_tick_child
+					{
+						// no child migration at hand, but one will begin here.
+						let maybe_first_child_key = {
+							let child_root = Pallet::<T>::child_io_key(top_key);
+							sp_io::default_child_storage::next_key(child_root, &vec![])
```

Comment:
I was told by emeric that this can only happen if someone uses the non-default child-tree api, which is currently not supported, and halt the migration if it happens.